be/src/util/fixed_hash_map.h

```diff
@@ -0,0 +1,125 @@
+    using iterator = iterator;
+
+    SmallFixedSizeHashMap() {
+        memset(_hash_table, 0, sizeof(ValueType) * hash_table_size);
+        _hash_table[hash_table_size] = reinterpret_cast<ValueType>(0xFFFF);
+    }
+
+    struct PPair {
+        using Cell = std::pair<KeyType, ValueType>;
+        PPair(KeyType key, ValueType value) : _data(key, value) {}
+        Cell _data;
+        Cell* operator->() { return &_data; }
+    };
+
+    class iterator {
+    public:
+        iterator(ValueType* hash_table_begin, uint32_t hash_table_key)
+                : _hash_table_key(hash_table_key), _hash_table_begin(hash_table_begin) {}
+
+        PPair operator->() const { return {static_cast<KeyType>(_hash_table_key), _hash_table_begin[_hash_table_key]}; }
+
+        iterator& operator++() {
+            _hash_table_key++;
+            skip_empty_value();
+            return *this;
+        }
+
+        void skip_empty_value() {
+            while (_hash_table_begin[_hash_table_key] == nullptr) {
+                ++_hash_table_key;
+            }
+        }
+
+        friend bool operator==(const iterator& a, const iterator& b) { return a._hash_table_key == b._hash_table_key; }
+        friend bool operator!=(const iterator& a, const iterator& b) { return !(a == b); }
+
+    private:
+        uint32_t _hash_table_key;
+        ValueType* _hash_table_begin;
+    };
+
+    template <class F>
+    iterator lazy_emplace(KeyType key, F&& f) {
+        auto search_key = static_cast<search_key_type>(key);
+        if (_hash_table[search_key] == nullptr) {
+            _size++;
+            f([&](KeyType key, ValueType value) {
+                DCHECK(value != nullptr);
+                _hash_table[search_key] = value;
+            });
+        }
+        return iterator(_hash_table, search_key);
+    }
+
+    iterator find(KeyType key) {
+        auto search_key = static_cast<search_key_type>(key);
+        if (_hash_table[search_key] != nullptr) {
+            return end();
+        }
+        return iterator(_hash_table, search_key);
+    }
+
+    iterator begin() {
+        auto iter = iterator(_hash_table, 0);
+        iter.skip_empty_value();
+        return iter;
+    }
+
+    iterator end() { return iterator(_hash_table, hash_table_size); }
+
+    void prefetch_hash(size_t hashval) const {
+        CHECK(false) << "unreachable path in fixed hash map";
+        __builtin_unreachable();
+    }
+
+    template <class F>
+    iterator lazy_emplace_with_hash(KeyType key, size_t& hashval, F&& f) {
+        CHECK(false) << "unreachable path in fixed hash map";
+        __builtin_unreachable();
+    }
+
+    std::hash<KeyType> hash_function() { return std::hash<KeyType>(); }
+
+    size_t bucket_count() { return hash_table_size; }
+
+    size_t size() { return _size; }
+
+    size_t capacity() { return bucket_count(); }
+
+    size_t dump_bound() { return hash_table_size; }
+
+private:
+    size_t _size = 0;
+    ValueType _hash_table[hash_table_size + 1];
```

Comment:
why size of array is hash_table_size + 1 not hash_table_size?

Reply:
in iterator::skip_empty_value
```c++
        void skip_empty_value() {
            while (_hash_table_begin[_hash_table_key] == nullptr) {
                ++_hash_table_key;
            }
        }
```
I init _hash_table[hash_table_size] to not-null value. so I don't have to determine the hash_key exceeds the length

```c++
        void skip_empty_value() {
            while (_hash_table_key < hash_table_size && _hash_table_begin[_hash_table_key] == nullptr) {
                ++_hash_table_key;
            }
        }
```